<a href="https://colab.research.google.com/github/cedeerwe/brutalna-akademia/blob/master/notebooks/05_odhadovanie.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Pomocný kód

Neotvárajte, obsahuje časti riešení.

In [0]:
import pandas as pd
import io 
import numpy as np

def get_age_data():
    csv_data = io.StringIO("""
Ukazovate�;;;1996;1997;1998;1999;2000;2001;2002;2003;2004;2005;2006;2007;2008;2009;2010;2011;2012;2013;2014;2015;2016;2017;2018
;;;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu;Spolu
Stav trvale b�vaj�ceho obyvate�stva k 31.12. (Osoba);Slovensk� republika;Spolu;5378932;5387650;5393382;5398657;5402547;5378951;5379161;5380053;5384822;5389180;5393637;5400998;5412254;5424925;5435273;5404322;5410836;5415949;5421349;5426252;5435343;5443120;5450421
;;Nula rokov;59594;58675;57157;55831;54741;50861;50519;51384;53451;54110;53637;54194;57151;60981;60164;60598;55828;55175;55513;56110;58277;58735;58445
;;1 rok;60762;59516;58586;57064;55758;53924;50805;50459;51355;53386;54086;53627;54221;57167;60952;57712;61091;56389;55747;56119;56967;59054;59484
;;2 roky;65623;60738;59499;58574;57043;55632;53919;50804;50458;51353;53383;54108;53648;54231;57189;59722;57903;61242;56576;55942;56394;57279;59358
;;3 roky;72487;65612;60693;59466;58556;56386;55635;53905;50803;50472;51370;53398;54121;53659;54270;57077;59805;58025;61365;56678;56156;56604;57499
;;4 roky;73665;72483;65578;60695;59464;57845;56395;55651;53914;50822;50469;51390;53409;54138;53663;54439;57141;59825;58073;61426;56811;56274;56746
;;5 rokov;77423;73647;72488;65569;60697;59967;57858;56403;55642;53936;50830;50469;51412;53423;54149;54008;54464;57122;59854;58113;61481;56905;56360
;;6 rokov;77329;77413;73636;72488;65552;61038;59961;57863;56400;55644;53939;50842;50472;51403;53432;54527;54015;54485;57082;59818;58108;61494;56950
;;7 rokov;77498;77335;77403;73636;72476;65833;61035;59970;57878;56407;55631;53941;50845;50485;51403;54020;54516;53958;54460;57040;59836;58151;61526
;;8 rokov;80736;77493;77330;77399;73644;72234;65821;61047;59968;57896;56413;55636;53950;50842;50483;51694;54003;54528;53935;54444;57021;59839;58156
;;9 rokov;81383;80726;77480;77316;77406;73797;72213;65840;61057;59970;57901;56415;55636;53960;50835;50915;51673;54024;54502;53922;54440;57009;59830
;;10 rokov;84299;81377;80720;77477;77308;77316;73778;72221;65833;61073;59972;57904;56418;55639;53962;51547;50889;51681;53991;54471;53908;54445;57002
;;11 rokov;87553;84296;81365;80707;77464;78778;77319;73788;72229;65842;61075;59976;57916;56421;55638;55037;51540;50879;51656;53965;54458;53903;54438
;;12 rokov;88112;87546;84278;81361;80718;79019;78780;77324;73798;72246;65841;61066;59974;57920;56430;55943;55028;51567;50857;51634;53955;54452;53892
;;13 rokov;88991;88102;87531;84259;81356;81940;79016;78783;77335;73805;72256;65832;61061;59982;57911;56763;55925;55048;51541;50821;51614;53941;54435
;;14 rokov;89442;88991;88097;87532;84242;82400;81937;79014;78794;77346;73819;72246;65835;61069;59976;58570;56755;55977;55029;51540;50802;51616;53921
;;15 rokov;89908;89460;88995;88110;87528;86016;82390;81947;79012;78796;77343;73828;72255;65836;61045;59618;58546;56848;55938;54989;51526;50796;51608
;;16 rokov;91492;89904;89464;88991;88105;88661;86003;82385;81936;79030;78799;77330;73820;72260;65836;60945;59596;58573;56824;55918;54969;51522;50792
;;17 rokov;96177;91458;89882;89443;88960;88771;88622;85994;82384;81918;79025;78799;77330;73826;72262;65684;60930;59702;58552;56806;55899;54963;51517
;;18 rokov;95853;96138;91409;89852;89406;89251;88726;88596;85982;82386;81919;79105;78861;77353;73852;72237;65680;60988;59685;58544;56816;55888;54960
;;19 rokov;95458;95801;96081;91370;89816;90582;89202;88681;88575;85976;82387;82064;79246;78911;77385;73411;72240;65797;60991;59712;58544;56814;55900
;;20 rokov;95166;95411;95742;96043;91321;90787;90531;89154;88659;88569;85989;82551;82241;79302;78938;76873;73419;72344;65799;61026;59749;58543;56810
;;21 rokov;93600;95095;95349;95673;95984;92154;90750;90461;89127;88661;88616;86133;82726;82320;79334;78299;76875;73547;72349;65825;61072;59764;58575
;;22 rokov;92999;93551;95050;95291;95601;96113;92068;90701;90463;89130;88707;88770;86280;82820;82355;78366;78320;77030;73557;72393;65865;61087;59798
;;23 rokov;88274;92947;93502;95014;95213;95152;96023;92030;90689;90489;89172;88871;88961;86368;82876;80907;78382;78465;77053;73599;72421;65895;61111
;;24 rokov;83445;88257;92880;93423;94965;94433;95042;95951;91999;90680;90548;89349;89040;89049;86424;81759;80929;78459;78487;77122;73647;72493;65943
;;25 rokov;78731;83418;88207;92858;93378;94647;94358;94984;95932;91991;90712;90680;89461;89130;89092;84175;81756;80948;78465;78506;77146;73678;72495
;;26 rokov;76250;78725;83408;88161;92777;92123;94553;94292;94949;95965;92052;90818;90812;89521;89214;87339;84172;81765;80927;78488;78547;77158;73738
;;27 rokov;74300;76219;78694;83346;88114;91704;92052;94495;94284;94941;95967;92196;90986;90861;89589;87476;87324;84055;81737;80962;78501;78559;77169
;;28 rokov;71048;74293;76213;78674;83298;86938;91617;91969;94475;94297;94938;96083;92350;91056;90888;87808;87456;87131;84035;81752;80987;78484;78513
;;29 rokov;71713;71032;74263;76181;78610;82191;86864;91574;91938;94446;94332;95062;96216;92399;91069;88266;87785;87188;87020;84009;81726;80996;78470
;;30 rokov;74788;71692;70996;74241;76123;77077;82131;86819;91589;91933;94467;94439;95144;96282;92412;88282;88252;87522;87082;86971;83990;81680;80942
;;31 rokov;77162;74731;71656;70967;74209;74586;77009;82079;86816;91566;91958;94558;94542;95180;96315;89607;88223;87840;87415;86993;86924;83959;81639
;;32 rokov;79701;77153;74683;71615;70924;73352;74552;76980;82087;86795;91597;92068;94652;94589;95206;93930;89574;88031;87714;87328;86930;86847;83907
;;33 rokov;79098;79650;77083;74634;71548;69899;73305;74524;76974;82089;86794;91698;92181;94677;94580;93099;93876;89088;87894;87653;87247;86865;86776
;;34 rokov;76229;79060;79572;77032;74581;70647;69837;73278;74534;76957;82085;86841;91757;92197;94685;92698;93036;93440;88894;87803;87528;87140;86772
;;35 rokov;78893;76174;78989;79509;76951;73999;70566;69788;73276;74538;76961;82155;86906;91768;92155;92212;92651;92699;93282;88730;87681;87431;87005
;;36 rokov;79536;78797;76080;78922;79455;76239;73941;70503;69743;73253;74562;77060;82217;86899;91777;90165;92136;92268;92524;93138;88623;87546;87317
;;37 rokov;78580;79425;78678;75999;78857;78691;76186;73890;70445;69733;73212;74624;77163;82210;86847;90133;90106;91915;92059;92360;92967;88477;87399
;;38 rokov;82528;78449;79321;78588;75908;77804;78586;76078;73846;70407;69714;73260;74668;77119;82196;85249;90066;89887;91738;91933;92189;92809;88348
;;39 rokov;85154;82383;78268;79166;78447;75292;77686;78462;76007;73803;70367;69795;73291;74651;77086;80361;85142;89835;89712;91612;91827;92057;92664
;;40 rokov;86806;84971;82196;78105;79036;77801;75166;77556;78372;75926;73744;70400;69843;73249;74610;75743;80281;85091;89625;89546;91441;91678;91891
;;41 rokov;86441;86640;84741;81993;77936;78142;77622;75028;77452;78276;75822;73697;70405;69776;73176;73460;75660;80147;84930;89455;89410;91268;91523
;;42 rokov;84774;86214;86405;84516;81821;77278;77931;77424;74916;77358;78186;75761;73696;70345;69703;72297;73379;75596;79980;84768;89284;89227;91087
;;43 rokov;84022;84522;85916;86154;84291;80762;77093;77739;77283;74805;77245;78101;75701;73562;70274;68808;72190;73316;75439;79854;84643;89131;89052
;;44 rokov;84172;83758;84235;85637;85880;83325;80522;76862;77564;77136;74636;77122;78015;75528;73410;69414;68720;72044;73152;75280;79683;84470;88946
;;45 rokov;82669;83887;83422;83937;85343;85331;83013;80223;76644;77346;76976;74509;77032;77868;75339;71993;69276;68577;71882;72973;75116;79502;84290
;;46 rokov;78799;82335;83528;83065;83619;84431;85032;82662;79938;76386;77127;76795;74364;76868;77674;74262;71798;69169;68362;71668;72792;74951;79314
;;47 rokov;71698;78418;81952;83188;82702;82627;84083;84708;82336;79638;76099;76891;76585;74161;76668;76806;74048;71575;68962;68168;71452;72569;74761
;;48 rokov;70789;71359;78006;81530;82793;81353;82211;83680;84353;81977;79309;75812;76634;76297;73892;75727;76545;73770;71346;68752;67968;71236;72349
;;49 rokov;67611;70400;70960;77577;81102;81928;80906;81767;83304;83953;81565;78971;75511;76336;76010;72787;75448;76265;73487;71099;68512;67721;70996
;;50 rokov;58416;67158;69928;70490;77101;80100;81418;80436;81358;82886;83571;81193;78609;75177;75974;75085;72467;75172;75940;73193;70847;68269;67491
;;51 rokov;53590;58026;66739;69422;70037;75973;79597;80906;79929;80918;82420;83114;80728;78181;74823;74962;74752;72087;74781;75623;72868;70500;67951
;;52 rokov;56019;53187;57582;66270;68904;69380;75435;79040;80430;79418;80381;81977;82592;80207;77724;73918;74547;74365;71731;74404;75278;72446;70135
;;53 rokov;52287;55555;52772;57134;65756;67985;68849;74860;78512;79844;78876;79858;81481;82040;79665;76891;73496;74153;73970;71288;74034;74863;72055
;;54 rokov;51733;51815;55030;52295;56625;64896;67414;68297;74269;77907;79233;78337;79299;80923;81461;78904;76354;72972;73693;73464;70842;73583;74390
;;55 rokov;51309;51246;51323;54459;51865;56011;64318;66821;67730;73645;77253;78559;77751;78687;80302;80184;78344;75835;72474;73173;72982;70368;73080
;;56 rokov;50874;50781;50704;50778;53914;51434;55421;63695;66275;67113;72995;76584;77890;77078;78042;79378;79543;77730;75225;71883;72632;72471;69841
;;57 rokov;47954;50286;50189;50159;50209;53426;50888;54778;63060;65603;66480;72255;75859;77139;76383;77168;78690;78881;77006;74537;71316;72055;71909
;;58 rokov;45990;47335;49651;49593;49587;49332;52830;50360;54208;62370;64912;65748;71476;75049;76472;75730;76491;77954;78123;76332;73886;70677;71433
;;59 rokov;44414;45364;46692;48987;48993;48934;48736;52160;49718;53583;61647;64193;64996;70645;74234;75438;74975;75708;77191;77349;75581;73173;70012
;;60 rokov;43570;43768;44720;46012;48322;48224;48250;48059;51465;49068;52907;60846;63410;64231;69776;73358;74543;74099;74869;76333;76559;74756;72387
;;61 rokov;44583;42877;43059;43994;45246;47206;47514;47549;47440;50729;48411;52221;60072;62550;63382;69052;72481;73578;73183;73934;75478;75593;73837
;;62 rokov;43662;43822;42134;42307;43279;44178;46449;46747;46847;46727;49987;47712;51481;59181;61637;62615;68110;71516;72577;72212;73008;74537;74675
;;63 rokov;43544;42830;42919;41349;41542;42016;43421;45659;45990;46121;46027;49209;46970;50640;58241;60743;61711;67128;70495;71480;71133;72001;73525
;;64 rokov;45540;42662;41943;42042;40556;40114;41226;42655;44844;45152;45344;45170;48374;46169;49802;57466;59717;60795;66133;69333;70390;70068;70962
;;65 rokov;44550;44558;41757;40995;41083;39476;39250;40379;41811;43979;44293;44525;44388;47522;45370;48661;56478;58765;59818;65047;68218;69276;68962
;;66 rokov;43796;43461;43452;40753;40009;39646;38563;38379;39531;40955;43073;43412;43700;43519;46566;44506;47794;55506;57789;58814;63905;67075;68042
;;67 rokov;42056;42614;42312;42304;39738;38711;38674;37641;37523;38587;40010;42109;42517;42816;42608;45690;43671;46927;54444;56708;57777;62834;65871
;;68 rokov;40680;40864;41334;41143;41157;38331;37704;37644;36673;36600;37593;39061;41129;41605;41889;41974;44727;42800;46022;53311;55549;56589;61547
;;69 rokov;38936;39408;39535;40101;39884;39607;37220;36603;36593;35666;35628;36531;38030;40144;40588;40943;41007;43765;41885;44953;52164;54367;55378
;;70 rokov;38226;37572;38045;38208;38779;38202;38427;36040;35447;35469;34655;34618;35462;36996;39152;39894;39943;40058;42710;40979;43890;51007;53154
;;71 rokov;36123;36774;36139;36690;36842;36880;36926;37138;34857;34313;34279;33598;33437;34419;35881;38419;38840;38875;39039;41591;39923;42740;49798
;;72 rokov;35522;34551;35268;34688;35218;35033;35480;35541;35795;33618;33097;33115;32471;32332;33298;35173;37350;37766;37849;37998;40491;38811;41584
;;73 rokov;36008;33872;33059;33797;33211;33291;33604;34055;34134;34368;32291;31865;31876;31280;31154;32604;34049;36204;36668;36731;36822;39291;37699
;;74 rokov;33759;34227;32155;31488;32204;31360;31754;32148;32532;32651;32889;30969;30587;30656;30143;30609;31485;32841;35051;35465;35630;35580;38026
;;75 rokov;31758;31940;32347;30459;29773;30288;29860;30178;30567;30976;31026;31397;29557;29298;29342;28995;29375;30315;31615;33796;34210;34425;34374
;;76 rokov;25319;29840;30033;30433;28733;27840;28654;28326;28618;28958;29365;29432;29896;28206;27865;28382;27732;28148;29072;30362;32464;32896;33094
;;77 rokov;23033;23696;27861;28148;28514;26760;26215;26909;26681;26983;27333;27734;27837;28311;26705;26885;26999;26564;26863;27744;29061;31067;31485
;;78 rokov;10207;21429;22047;25933;26240;26273;24985;24401;25129;24966;25206;25591;26034;26242;26637;25571;25437;25714;25257;25543;26463;27664;29576
;;79 rokov;9654;9409;19777;20319;24004;24199;24368;23280;22609;23355;23206;23437;23922;24347;24558;25274;24090;23960;24338;23804;24158;25025;26216
;;80 rokov;9615;8845;8575;18128;18566;21807;22234;22374;21425;20789;21537;21387;21714;22219;22558;23243;23622;22588;22429;22817;22478;22796;23615
;;81 rokov;11876;8734;8048;7773;16420;16703;19899;20256;20426;19467;19016;19666;19672;20001;20294;21213;21514;21948;21032;20811;21288;21002;21248
;;82 rokov;15357;10679;7857;7291;6953;14613;15049;18044;18376;18401;17630;17150;17825;17895;18167;18870;19435;19722;20169;19369;19237;19666;19448
;;83 rokov;13219;13711;9444;6950;6432;6009;13051;13454;16138;16447;16396;15743;15486;15969;16101;16668;17057;17741;18049;18371;17771;17581;18072
;;84 rokov;11667;11614;12013;8257;6093;5506;5323;11468;11861;14152;14570;14418;13908;13722;14107;14667;14865;15346;16047;16276;16713;16098;15923
;;85 rokov;9239;10199;10111;10534;7201;5236;4847;4626;10013;10308;12382;12685;12615;12163;11977;12700;12950;13212;13824;14296;14601;14986;14424
;;86 rokov;8040;7901;8710;8678;9041;6056;4477;4170;3998;8549;8823;10693;10887;10928;10479;10621;11070;11308;11627;12099;12637;12855;13259
;;87 rokov;7139;6785;6624;7348;7313;7413;5152;3788;3560;3332;7201;7425;9117;9289;9360;9218;9021;9623;9827;10073;10531;10977;11145
;;88 rokov;5451;5905;5595;5501;6149;5813;6105;4302;3203;2975;2755;5984;6232;7621;7710;8186;7832;7784;8185;8344;8624;9079;9455
;;89 rokov;3948;4412;4787;4576;4439;4867;4756;5012;3528;2606;2434;2290;4903;5113;6267;6484;6779;6658;6608;6843;7036;7177;7681
;;90 rokov;3120;3160;3535;3865;3687;3518;3940;3788;4034;2814;2103;1944;1898;3901;4089;5107;5310;5539;5506;5512;5673;5797;5963
;;91 rokov;2300;2465;2511;2825;3040;2754;2756;3088;2985;3167;2231;1654;1526;1493;3067;3324;4075;4310;4475;4476;4523;4582;4707
;;92 rokov;1808;1786;1948;1964;2199;2076;2157;2134;2407;2260;2464;1735;1301;1186;1164;2541;2580;3225;3477;3505;3611;3594;3643
;;93 rokov;1340;1418;1448;1514;1563;1418;1597;1619;1638;1807;1709;1913;1333;960;904;953;1964;2008;2532;2729;2725;2843;2865
;;94 rokov;992;1028;1139;1162;1246;947;1055;1177;1234;1238;1340;1267;1423;1022;740;764;723;1497;1545;1934;2070;2075;2230
;;95 rokov;688;776;839;898;917;632;654;744;870;912;928;1009;962;1059;776;585;576;561;1155;1182;1461;1543;1603
;;96 rokov;464;555;634;689;740;504;448;479;544;629;652;689;747;723;782;552;450;446;449;847;927;1095;1216
;;97 rokov;280;383;473;561;582;326;363;330;325;394;459;484;497;572;560;476;398;338;350;330;640;724;875
;;98 rokov;192;234;311;411;485;188;228;259;238;236;289;355;363;382;464;330;341;300;287;279;265;502;585
;;99 rokov;132;172;200;265;379;153;123;170;195;171;167;234;262;286;312;259;239;243;245;227;231;222;407
;;100 rokov;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;152;193;184;186;199;185;198;190
;;100 rokov alebo viac;192;284;407;538;750;249;327;367;451;540;609;663;791;937;1090;0;0;0;0;0;0;0;0
;;101 rokov;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;60;119;161;151;153;174;156;182
;;102 rokov;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;36;44;99;136;133;142;149;140
;;103 rokov;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;27;26;31;94;121;120;132;142
;;104 rokov;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;14;21;16;26;92;114;113;128
;;105 rokov;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;6;11;13;11;24;88;108;106
;;106 rokov;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;9;5;6;11;10;23;86;105
;;107 rokov;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;8;8;5;6;12;9;24;84
;;108 rokov;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;4;7;6;4;6;11;11;24
;;109 rokov;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;2;4;4;6;4;6;11;10
;;110 rokov alebo viac;0;0;0;0;0;0;0;0;0;0;0;0;0;0;0;3;6;6;10;16;20;26;38
""")

    data = pd.read_csv(csv_data, delimiter=";")
    data = data[data.columns[2:]].iloc[1:, :]
    index = data.iloc[:, 0].values
    data = data[data.columns[1:]]
    data.index = index
    return data.applymap(int)

# Úloha č.1

Na obrázkoch vidíte krabicu s kačičkami.


<a href="https://ibb.co/2nkKDrm"><img src="https://i.ibb.co/2nkKDrm/IMG-20200107-202200.jpg" alt="IMG-20200107-202200" border="0" /></a>
<a href="https://ibb.co/vsNkvmV"><img src="https://i.ibb.co/vsNkvmV/IMG-20200107-202121.jpg" alt="IMG-20200107-202121" border="0" /></a>
<a href="https://ibb.co/FbCgxLW"><img src="https://i.ibb.co/FbCgxLW/IMG-20200107-202129.jpg" alt="IMG-20200107-202129" border="0" /></a>
<a href="https://ibb.co/g3KRmpg"><img src="https://i.ibb.co/g3KRmpg/IMG-20200107-202140.jpg" alt="IMG-20200107-202140" border="0" /></a>
<a href="https://ibb.co/86xJJjZ"><img src="https://i.ibb.co/86xJJjZ/IMG-20200107-202148.jpg" alt="IMG-20200107-202148" border="0" /></a>

1. Odhadnite čo najpresnejšie počet kačičiek v krabici.
1. Určte interval, ktorému na 95% veríte, že sa v ňom nachádza správny výsledok na časť 1. (Ak by som vám dal podobnú úlohu 20 krát, mali by ste sa mýliť vo svojom intervale v priemere raz.) 



## Riešenie

Takýto typ úlohy sa veľmi často vyskytuje v praktickom využití matematiky. Máme hodnotu, ktorú sa snažíme odhadnúť / modelovať, ale máme k dispozícii iba obmedzené množstvo informácií. V praxi to môžu byť výsledky volieb, počet predajov v našom obchode v ďalšom mesiaci, celkové náklady na výlet do Paríža, v princípe hocičo. V každom z týchto scenárov by sme odhadovanú hodnotu chceli použiť iným spôosobom.

Napríklad pri výlete do Paríža je pre nás dôležité zhora ohraničiť, koľko nás tento výlet vyjde, aby sme zistili, či na to máme peniaze. Pri počte predajov spotrebného tovaru by sme mohli chcieť vedieť dolné ohraničenie predaja, aby sme nekúpili viac a vyhli sme sa pokazenému tovaru. Vo volebných prieskumoch väčšinu ľudí zaujíma konkrétne čislo a žiadne ohraničenie. 

V niektorých prípadoch nám stačí **bodový odhad** (časť 1), ale vo veľa prípadoch sa nám viac hodí uvažovať ohraničenia tejto hodnoty, takzvaný **intervalový odhad** (časť 2). Interval je síce širší odhad, ale vieme priamo pracovať s pravdepodobnosťou jeho presnosti. V praxi je veľmi populárny. Viac informácií o intervalovom odhade (resp. intervale spoľahlivosti) sa môžte dočítať [tu](https://en.wikipedia.org/wiki/Confidence_interval).

V úlohách takéhoto typu sa budeme stretávať s problémom spokojnosti. Nebude jasné, či máme správny výsledok a nebudeme sa o tom ani vedieť presvedčiť. V tejto úlohe ešte môžme manuálne spočítať kačičky (ak ich máme k dispozícii), ale pri ostatných vôbec nie je jasné, kedy by sme mali byť s úlohou škončený.

Budeme sa preto snažiť spraviť viacero rozličných odhadov a na záver si porovnáme ich všetky výsledky.

### Odhad 1


Náš prvý odhad spravíme nasledovne:
1. spočítame si všetky kačičky, ktoré sú jasne viditeľné na obrázku č.4.
1. odhadneme pomocou obrázku 2, akú časť všetkých kačičiek sme spočítali.
1. dopočítame priamou úmerov počet kačičiek v celej krabičke.

Opísali sme teraz metodiku nášho odhadu, skúsime si ho teraz spraviť. Spravte si ho sami predtým, ako ho spravíme za vás my.



#### Počítanie kačičiek


<a href="https://ibb.co/ZWWtd1S"><img src="https://i.ibb.co/MMMWn5f/kacicky-zvrchu.jpg" alt="kacicky-zvrchu" border="0" /></a>

Na obrázku sme zaznačili 33 kačičiek. Niektorým kačičkám je ešte vidieť hlava alebo chvost, ale tie sme teraz nepočítali.

#### Odhad zlomku

V tejto časti si budeme zaznačovať aj náš interval odhadu. Nie je totiž vôbec jasné, akú veľkú časť kačičiek sme spočítali. Pozrime sa na obrázok č. 2 a zakreslime si, pokiaľ siahajú naše vyznačené kačičky.

<a href="https://ibb.co/tLR06tY"><img src="https://i.ibb.co/GHmZDLC/kacicky-siahanie.jpg" alt="kacicky-siahanie" border="0" /></a>

Prvá čiara je horná hranica nášho odhadu, fialová dolná hranica a čeverná je náš bodový odhad. Pravdaže sa môžme mýliť, ale takto nám to od pohľadu vyšlo. Pomocou fialových čiar teraz dostaneme aj intervalový odhad.

Spočítajme si teraz v akej časti celej výšky krabičky sa tieto čiary nachádzajú. V grafickom programe sme si pozreli počet pixelov od vrchu krabice po spodok. Napočítali sme 930. Po prvú fialovú čiaru je 230 pixelov, po červenú 330 a po poslednú 410. 

Kačičky ale pravdepodobne budú hustejšie nakopené na spodku ako na vrchu. Aj z ostatných obrázkov môžme vidieť, že je na vrchu na niektorých miestach prázdno a kačičky moc nepresahujú výšku krabičky. Môžme teda upraviť výšku našej krabičky o niekoľko pixelov, aby sme toto zobrali v úvahu. Odpočítame teda od všetkých hodnôt 50 pixelov, čo sme si odhadli okom podľa grafického programu.

#### Odhad výsledku

V tejto časti iba dáme dokopy čísla, ktoré sme dostali. Robíme to separátne od prvých dvoch častí, aby sme napríklad nevolili čiary podľa toho, čo sa nám zdá ako dobrý výsledok.

Ak by sme nerátali s úpravou o prázdny priestor na vrchu krabičky, dostali by sme bodový odhad 

$$\tfrac{930}{330} \cdot 33 = 93$$ 

kačičiek s intervalom spoľahlivosti 

$$[\tfrac{930}{410} \cdot 33 \approx 74.85 , \tfrac{930}{230} \cdot 33 \approx 133.43].$$ 

So zapracovaním korekcie za prázdny priestor dostávame bodový odhad 

$$\tfrac{880}{280} \cdot 33 \approx 103.71$$ 

kačičiek s intervalom spoľahlivosti 

$$[\tfrac{880}{360} \cdot 33 \approx 80.66, \tfrac{880}{180} \cdot 33 \approx 161.33].$$

#### Zamyslenie sa nad výsledkom

Naše bodové odhady vyzerajú na prvý pohľad rozumne. Netušíme, či sú správne, ale rádovo sedia s našimi prvotnými pocitmi. Čo sa týka intervalov, spodné ohraničenia vyzerajú tiež rádovo okej, ale vrchné vyzerá byť ustrelené, hlavne v druhom prípade. Keď sa nad tým zamyslíme, po korekcii o 50 pixelov sa naša vrchná fialová čiara veľmi priblížila k vrchu krabičky a teda výpočet $\tfrac{880}{180}$ očividne nezodpovedá nášmu reálnemu odhadu. Mali by sme prekresliť naše čiary s tým, že rátame s korekciou od začiatku. Získanie lepšieho intervalu týmto spôsobom už necháme na čitateľa.

Intervaly spoľahlivosti sú často veľmi široké, ako si môžme všimnúť aj v tomto príklade. Sme si *dosť istý* tým, že skutočný počet kačičiek sa naozaj v tomto intervale bude nachádzať, takže úloha je splnená. Pre praktické využitie je najlepšie, keď sú intervaly čo najužšie. Keby naša vláda odhadla, že na Slovensku aktuálne žije medzi 100 tisíc a 100 miliónov ľudí, ťažko by sa jej robil štátny rozpočet. Preto je dôležité spraviť tak úzky interval ako to ide, ale aby sme zároveň nestratili našu silnú istotu (95%), že je interval správny. 

Ohľadom čísla 95%: v praxi sa veľmi často používa 95%. Vyjadruje istú mieru pochybnosti, ktorá sa dá schovať do rozumne úzkeho intervalu a zároveň vyjadruje dostatočnú istotu, nad ktorou sa dá budovať. Vo väčšine prípadov nikto netuší, či sa skutočne pohybujeme okolo 95% - to je ale v poriadku. Dôležité je, že si všetci uvedomujú málopercentnú šancu na chybu a veľkopercentnú istotu správneho ohraničenia. V niektorých situáciách sa vyžaduje aj väčšia spoľahlivosť, keď riziko nesprávneho výsledku môže znamenať katastrofu. Vtedy sa môžu všetci oháňať číslom napríklad 99.9% a omnoho širšími intervalmi.

Posledná poznámka je k bodovému odhadu 103.71. Dáva nám zmysel odhadnúť neceločíselné riešenie, keď je počet kačičiek určite celočíselný? Zamyslite sa a prediskutujte.

### Odhad 2


Aby sme dostali čo najinformatívnejší druhý odhad, mala by sa jeho metodika čo najviac líšiť od predošlého. Takýmto spôsobom zaručíme, že naše odhady budú čo najviac nezávislé a teda znížime šancu tomu, že spravíme dvakrát takú istú chybu.

Spravíme teda odhad na základe objemu. Odhadneme objem krabice, objem každej kačičky a spočítame, koľko by sa ich tam zmestilo. Pravdaže, v krabičke je aj veľa vzduchu, na čo budeme musieť myslieť.

Metodika teda bude vyzerať takto:
1. Odhadneme objem krabičky
1. Odhadneme objem kačičky
1. Odhadneme počet kačičiek v krabičke

Opäť by sme vás radi vyzvali, aby ste si tieto kroky skúsili spraviť sami.

#### Odhad veľkosti krabičky


Použijeme obrázok č. 4 na vypočítanie obsahu prierezu a následne obrázok č.5 na spočítanie výšky krabičky. 

Na obrázku č. 4 sme spočítali nasledovné rozmery:

<a href="https://ibb.co/0Z4X9VG"><img src="https://i.ibb.co/6b9ZtJW/krabicka-rozmery-vrch.jpg" alt="krabicka-rozmery-vrch" border="0" /></a>

Celá dĺžka (horizontálna) 3880 pixelov, šírka (vertikálna) 2300 pixelov a dĺžka medzi oblúkovými časťami 1700 pixelov. Oblúkové časti budeme považovať za kružnice. Máme teda dva poloblúky, ktoré dokopy vytvárajú kruh o priemere 2600 pixelov. Obsah tohto kruhu je teda $1150^2 \cdot \pi \approx 4\,154\,756$ pixelov štvorcových. Obdĺžniková časť medzi polkruhami má obsah $2300 \cdot 1700 = 3\,910\,000$ pixelov štvorcových. Dokopy je to $8\,064\,756~\text{px}^2$.

Teraz skúsime vyrátať výšku krabičky. Na obrázku č. 5 sme zrátali nasledovné rozmery:

<a href="https://ibb.co/17bVX3F"><img src="https://i.ibb.co/DzLXMN0/krabicka-rozmery-zboku.jpg" alt="krabicka-rozmery-zboku" border="0" /></a>

Horizontálna dĺžka 2960 pixelov a výška 1530 pixelov. Treba si dávať pozor na perspektívu a uvedomiť si, ktoré z čiar treba naozaj merať. Musíme si tiež dávať pozor na jednotky. Pixel na prvom obrázku má inú veľkosť ako pixel na druhom obrázku. Keďže budeme merať veľkosti kačičiek podľa pixelov na prvom obrázku, prevedieme si pixely z druhého obrázka na pixely prvého obrázka. Na to nám slúži dĺžka krabičky. Dĺžka bola na prvom obrázku 3880 pixelov a na druhom 2960 pixelov. Výška krabičky v jednotkách prvého obrázka bude $\tfrac{3880}{2960} \cdot 1530 \approx 2005$ pixelov.

Celý objem krabičky sme vyrátali ako 
$$
8\,064\,756 ~\text{px}^2 \cdot 2005 ~\text{px} = 16\,169\,835\,780 ~\text{px}^{3}
$$

To je celkom dosť 3D pixelov, ktoré sa zmestia do tejto krabičky.

#### Odhad veľkosti kačičky



Na prvom obrázku sme zmerali potrebné rozmery kačičky na viacerých príkladoch a zobrali sme priemerné číslo.

<a href="https://ibb.co/H7N1K1V"><img src="https://i.ibb.co/TcKpwpk/kacicky-rozmery.jpg" alt="kacicky-rozmery" border="0" /></a>

Dostali sme sa k číslam dĺžka 565 pixelov, šírka 470 pixelov (bez krídel, ktoré budeme ignorovať) a výška 520 pixelov. Náročná otázka ale je, ako budeme modelovať objem kačičky. Je zjavné, že nie je všetok priestor krabičky vyplnený kačičkami. Horný odhad spravíme ako kváder, teda 

$$
565 ~\text{px} \cdot 470 ~ \text{px} \cdot 520~\text{px} = 138\,086\,000 ~\text{px}^3.
$$

Pre tento odhad platí, že sme si celkom istý, že kačička aj so vzduchom nebude viac ako takýto kváder. 

Čo sa týka spodného odhadu, môžme použiť ihlan rovnakej výšky, ale je celkom jasné, že samotná kačička bude mať aspoň taký objem (bez vzduchu). Takýto ihlan by mal tretinu objemu kvádra, čo nám príde ako veľmi slabý odhad. Spravíme teda spodný odhad kačičky so vzduchom ako polovicu kvádra. Okom to vyzerá, že by to naozaj mohol byť spodný odhad, v ktorom sme si celkom istý. Náš spodný odhad je teda $69\,043\,000 ~\text{px}^3$. 

Ako skutočný odhad použijeme číslo presne medzi našimi odhadmi, teda $\tfrac{3}{4}$ kvádra. Od pohľadu nevieme spraviť lepší odhad. Dalo by sa argumentovať aj za $\tfrac{2}{3}$. Bohužiaľ, odhadnúť presný objem kačičky aj so vzduchom, ktorý zaberá, je veľmi náročné a bude to zjavne nepresné. Naše spodné a vrchné odhady by ale mali zachytiť výsledky z minulého odhadu, ak im naozaj veríme. Bodový odhad teda bude $103\,564\,500 ~\text{px}^3$.

#### Odhad výsledku

Teraz nám ostáva spraviť záverečné výpočty. Je dôležité, že ich nerobíme priebežne s určovaním objemu kačičky, pretože by sme boli ovplyvnený výsledkami a odhad by bol závislý od nášho prvého odhadu.

Dostávame bodový výsledok

$$
\frac{16\,169\,835\,780}{103\,564\,500} \approx 156.13
$$

a interval ako 

$$
\left[ \frac{16\,169\,835\,780}{138\,086\,000} \approx  117.1 , \frac{16\,169\,835\,780}{69\,043\,000} \approx  234.2 \right].
$$

#### Zamyslenie sa nad výsledkom

Výsledky nám vyšli celkom výrazne vyššie ako v predchádzajúcom prístupe. Dokonca aj spodný interval odhadu, kde sme kačičku odhadovali plným kvádrom, nám vyšiel viac ako bodový odhad v predchádzajúcom prístupe.

Zamyslite sa, čo sme mohli spraviť zle v týchto prístupoch. Ktorý výsledok by sme mali preferovať? Prečo? Na základe týchto dvoch odhadov, ký by sme si mali tipnúť konečný výsledok?

### Odhad 3

Keď sa dve metodiky natoľko líšia vo výsledkoch, je dobrý nápad skúsiť spraviť ešte tretí odhad. Ak ste tak ešte nespravili, navrhnite aspoň jednu ďalšiu metodiku a vypočítajte na jej základe odhad.

### Výsledok



Ostáva nám spočítať kačičky.

<a href="https://ibb.co/tq7Cxv4"><img src="https://i.ibb.co/D9xK8vt/IMG-20200107-202750.jpg" alt="IMG-20200107-202750" border="0" /></a>

Je ich dokopy 102.

#### Ponaučenia

Iste ste si všimli dvoch kačičkových *generálov*, ktorý sú výrazne väčší, ako všetkých 100 zvyšných kačičiek. Na základe našich obmedzených dát sme chybne predpokladali, že majú všetky kačičky rovnakú veľkosť. Bol to dobrý predpoklad na základe našich informácií, ale nikdy sme na to nemali záruku. Takto to bohužiaľ býva aj v reálnom svete. *Naše dáta nemusia byť nikdy dokonale výpovedné*.

Takisto si pozornejší z vás možno všimli, že jedna z kačičiek mala na sebe zo spodu napísané číslo 103, čo napovedalo tomu, že ich môže byť viac ako 103. Vaše odhady sa na základe tejto indície možno podvedome pohybovali výhradne nad týmto číslo. O pôvode týchto čisel sme nemali žiadnu informáciu (aj keď je pravda, že kačičky boli očíslované v poradí - neboli ale všetky použité). Je dôležité vediet odfiltrovať spoľahlivé informácie od nespoľahlivých. *Nie všetky informácie sú nutne relevantné.*

Niektoré naše bodové odhady boli naozaj mimo. To sa stáva. Preto sú v praxi tak populárne intervalové odhady, pretože aspoň dostávame rozumnú záruku na počet. Bodový odhad môže byť veľmi mimo. Interval môže byť síce široký, ale dáva nám celú informáciu typu: "Výsledok je z nášho pohľadu na 95% v tomto intervale a nevieme ho zúžiť, lebo je možné, že je výsledok blízko jeho hraníc". Hovorí to niečo o výsledku, ktorý hľadáme. Je neistý. *Neistota sa dá veľmi prakticky zachytiť intervalom*.

## Úloha č.2

Koľko občanov Slovenska sa tento rok stane plnoletými? 

Spravte bodový odhad (časť 1 predchádzajúcej úlohy) aj 95% odhad intervalu (časť 2 predchádzajúcej úlohy).

## Riešenie

Na rozdiel od predchádzajúcej časti nemáme žiadne dáta k dispozícii, budeme si ich musieť nájsť. Zatiaľ nám nie je jasné, na základe akého modelu budeme budovať náš odhad, lebo nevieme, čo máme k dispozícií. Prvý krok je teda jasný, ideme sa prehrabať internetom.


### Hľadanie dát

Nachádzanie správnych a relevantných dát je výzvou. Použijeme na to určite google, otázkou ostáva, aké pojmy budeme googliť. 





#### Prvý pokus


Na prvý pokus sme vyskúšali hľadať "populacia slovensko podla veku". Medzi prvými odkazmi sme našli [tento](http://obyvatelstvo.population.city/slovensko/). Tvári sa, že má aktuálne štatistiky počtu obyvateľska v rôznych vekových skupinách. Naša záujmová skupina sa bude nachádzať v kategórii 15-19. Náš prvý odhad teda môže byť počet mužov + počet žien v tejto kategórii, vydelené piatimi - lebo nás zaujíma iba jeden z piatich vekov v tejto kategórii. Náš prvý bodový odhad je teda 

$$\frac{122\,579 + 116\,980}{5} = 47911.8.$$ 

Nevieme ale odkiaľ presne tieto dáta sú a či sa na ne dá spoľahnúť. Na stránke sa dá prekliknuť ku zdrojom, čo nás prenesie na stránku OSN. Po rozkliknutí tabuliek *Population by age* a navolení niekoľkých základných údajov sa vieme dostať k číslu $263\,000$ ako k súčtu mužov a žien v roku 2020 medzi 15 a 19. To nás vedie k druhému odhadu

$$\frac{263\,000}{5} = 52600.$$

Stránke OSN asi môžme celkom veriť. Radi by sme dostali ale nejakú presnejšiu štatistiku, pretože v skupine 15-19 sa nemusia vyskytovať 18-ročný rovnomerne s ostatnými. Mohli by sme využiť historické dáta týchto skupín a odhadnúť pomery 15, 16, 17, 18 a 19 ročných v tejto skupine cez ich prechody medzi iné skupiny (premyslite si, vyskúšajte). Pre teraz ale skúsime nájsť relevantnejšie dáta.

#### Druhý pokus

Vyskúšali sme preformulovanie: "Vekové rozloženie obyvateľstva slovensko". Prvý odkaz vyzerá veľmi sľubne, prišli sme na [túto stránku](http://statdat.statistics.sk/cognosext/cgi-bin/cognos.cgi?b_action=cognosViewer&ui.action=run&ui.object=storeID%28%22i40A03AF2150C41DE8BE98D0C0C41A764%22%29&ui.name=Vekov%C3%A9%20zlo%C5%BEenie%20-%20SR%2C%20oblasti%2C%20kraje%2C%20okresy%2C%20mesto%2C%20vidiek%20%5Bom7009rr%5D&run.outputFormat=&run.prompt=true&cv.header=false&ui.backURL=%2Fcognosext%2Fcps4%2Fportlets%2Fcommon%2Fclose.html&run.outputLocale=sk). Skvelé, to je presne to, čo sme potrebovali. V ľavom stĺpci sa nám ukazuje *Stav trvale bývajúceho obyvateľstva k 30.6.(1.7.) (Osoba)* respektíve *Stav trvale bývajúceho obyvateľstva k 31.12. (Osoba)*. Nás by zaujímalo druhé z týchto čísel, keďže počet 17 ročných 31.12.2019 v roku 2019 by mohol veľmi blízko zodpovedať počtu obyvateľov nadobúdajúcich plnoletosť v roku 2020. Na spodku stránky si môžme všimnúť poznámku, že stav v strede roka sa počíta ako aritmetický priemer najbližších koncov roka a teda nemá pre nás až tak silnú výpovednú hodnotu.

Zdroj vyzerá byť veľmi spoľahlivý, keďže ide o štatistický úrad Slovenskej Republiky. Všetky tieto údaje by sme mali brať ako najpresnejšie odhady, aké vie Slovensko spraviť. Budeme sa na ne spoliehať.

Bohužiaľ, údaje o roku 2019 nie sú k dispozícií. Keďže posledná aktualizácia bola 10.4.2019, môžme predpokladať, že niekedy v apríli tohto roka sa stane nasledujúca aktualizácia. Budeme teda musieť pracovať s rok starými dátami. To nie je problém, často musíme v praxi pracovať s omnoho staršími dátami a "predpovedať" ich správanie do budúcnosti (predpovedaniu na základe historických dát sa vo vedeckom svete hovorí **extrapolácia**). Ak ste sa k týmto dátam počas svojho hľadania nedostali, dajte si chvíľku na predpovedanie výsledku na základe tejto tabuľky. 

### Extrapolácia

Po preklikaní niektorých nastavení sme zo stránky v predchádzajúcom odseku získali nasledovnú tabuľku:

In [0]:
data = get_age_data() # vyčistí a spracuje data, ktoré ste si pozreli na webe
relevant_data = data.iloc[1:20,:] # ukáže prvý až dvadsiaty riadok dát
relevant_data

1996   1997   1998   1999  ...   2015   2016   2017   2018
Nula rokov  59594  58675  57157  55831  ...  56110  58277  58735  58445
1 rok       60762  59516  58586  57064  ...  56119  56967  59054  59484
2 roky      65623  60738  59499  58574  ...  55942  56394  57279  59358
3 roky      72487  65612  60693  59466  ...  56678  56156  56604  57499
4 roky      73665  72483  65578  60695  ...  61426  56811  56274  56746
5 rokov     77423  73647  72488  65569  ...  58113  61481  56905  56360
6 rokov     77329  77413  73636  72488  ...  59818  58108  61494  56950
7 rokov     77498  77335  77403  73636  ...  57040  59836  58151  61526
8 rokov     80736  77493  77330  77399  ...  54444  57021  59839  58156
9 rokov     81383  80726  77480  77316  ...  53922  54440  57009  59830
10 rokov    84299  81377  80720  77477  ...  54471  53908  54445  57002
11 rokov    87553  84296  81365  80707  ...  53965  54458  53903  54438
12 rokov    88112  87546  84278  81361  ...  51634  53955  54452  53892
13 rokov    88991  88102  87531  84259  ...  50821  51614  53941  54435
14 rokov    89442  88991  88097  87532  ...  51540  50802  51616  53921
15 rokov    89908  89460  88995  88110  ...  54989  51526  50796  51608
16 rokov    91492  89904  89464  88991  ...  55918  54969  51522  50792
17 rokov    96177  91458  89882  89443  ...  56806  55899  54963  51517
18 rokov    95853  96138  91409  89852  ...  58544  56816  55888  54960

[19 rows x 23 columns]

Skúsime si tieto dáta aj vykresliť, aby sme v nich lepšie videli vzory. Vedieť správne vizualizovať (a vyčistiť) dáta je kľúčová schopnosť, ktorá je často medzi matematikmi podceňovaná.

In [0]:
import plotly.graph_objects as go
fig = go.Figure( go.Heatmap(
    x=relevant_data.columns, 
    y=relevant_data.index[::-1],
    z=relevant_data.values[::-1,:],
    colorscale="viridis"
    ) )
fig.show()

Je celkom jasne vidieť, že veľkosti skupín sa z roka na rok moc nemenia. Dobrý odhad na náš výsledok by teda bol počet 16 ročných z roku 2018, čo je 50792. Je to číslo niekde v strede medzi našimi prvými dvoma odhadmi, ktoré sme získali.

Keď sme sa už zoznámili s našími dátami, je na čase si skúsiť napísať, čo sa snažíme predpovedať a ako to budeme modelovať. Naším cieľom je predpovedať počet 17 ročných v roku 2019. Náš odhad bude založený na počte 16 ročných v roku 2018 a korekcii za pridaný rok. Za tento rok mohli niektorý zo 17 ročných odísť zo Slovenska,  umrieť, alebo sa dokonca prisťahovať. Túto korekciu sa budeme snažiť zachytiť na základe našich dát.

Prvý model, ktorý skúsime aplikovať je nasledovný. Budeme predpokladať, že naša skupina narodených v roku 2002 nie je ničím špeciálna a ich chovanie medzi 16tym a 17tym rokom bude zodpovedať chovaniu všetkých v minulosti v tejto skupine. Vypočítame si teda priemernú hodnotu 

$$
\frac{17~\text{ročný v roku}~(N + 1)}{16~\text{ročný v roku}~N}.
$$
Z našich dát to dostaneme nasledovne:

In [0]:
age16 = relevant_data.loc["16 rokov"].values[:-1] # všetci 16 ročný okrem posledného
age17 = relevant_data.loc["17 rokov"].values[1:] # všetci 17 ročný okrem prvého

pomery = np.divide(age17, age16)
pomery

array([0.99962838, 0.99975529, 0.99976527, 0.99965165, 1.00755916,
       0.99956012, 0.99989535, 0.99998786, 0.99978032, 0.99993673,
       1.        , 1.        , 1.00008128, 1.00002768, 0.99769123,
       0.99975388, 1.00177864, 0.99964147, 0.99968323, 0.99966022,
       0.99989085, 0.99990295])

Priemerný pomer teraz dostaneme vypočítaním priemeru týchto čísel:

In [0]:
np.mean(pomery)

1.00016507168112

Tento výsledok je kúsok prekvapivý. Keď sa pozrieme na pomery, ktoré nám vyšli, 16 z nich je pod 1, dva sú presne 1 a iba 4 sú väčšie ako 1. Tie štyri ale majú väčšiu váhu, ako všetky menšie ako 1 dokopy, takže sa žiadna chyba v matematike nestala. Vo väčšine prípadov ale počet 17 ročných je kúsok menší ako počet 16 ročných, takže ak predpovieme počet 17 ročných na základe tohto výsledného čísla, vo väčšine prípadov sa budeme mýliť.

Z tohto dôvodu sa v podobných situáciách často používa namiesto aritmetického priemeru *medián*. Medián získame tak, že si zoradíme všetky hodnoty a zoberieme tú v strede (v prípade párneho počtu hodnôt priemer dvoch stredných). Skúsme teda spraviť vypočítať medián týchto pomerov.

In [0]:
np.median(pomery)

0.9998355819565468

To je už uveriteľnejšie. Ak na základe tohto čísla predpovieme počet 17 ročných, dostaneme 

$$50792 \cdot 0.9998356 \approx 50783.65.$$

Takže iba na základe predpokladu, že skupina narodená roku 2002 nie je ničím odlišná od skupín narodených iné deti, dostali sme prvý prepracovaný odhad nášho výsledku. 

Ak by sme chceli spraviť ešte intervalový odhad týmto modelom, mohli by sme zobrať najmenší a najväčší pomer, ktorý sme za tie roky videli a použiť tie. Dostaneme takto interval

In [0]:
50792 * np.min(pomery), 50792 * np.max(pomery)

(50674.73309435567, 51175.944974746046)

#### Extrapolácia pokračovanie


Posledný krok je overiť, či je náš predpoklad správny - či skupina z roku 2002 je niečím špeciálna alebo nie. Môžme to overiť viacerými spôsobmi. Skúsime sa pozrieť na postupne na úbytky / prírastky medzi vekmi 0 a 1,  1 a 2, 2 a 3, 3 a 4, atď.. Ak sa vo všetkých týchto úbytkoch umiestni naša skupina 2002 ako priemerná, môžme začať lepšie veriť nášmu minulému odhadu. Spravíme to zase vizualizáciou, aj keď to teraz bude mierne komplikovanejšie.

In [0]:
results = {}
for i_vek in range(1, len(relevant_data.index)):
    for i_rok in range(1, len(relevant_data.columns)):
        pomer = relevant_data.iloc[i_vek, i_rok] / relevant_data.iloc[i_vek-1, i_rok-1]
        narodeni = int(relevant_data.columns[i_rok]) - i_vek
        if narodeni not in results:
            results[narodeni] = {'x': [], 'y': []}
        results[narodeni]['x'].append(i_vek)
        results[narodeni]['y'].append(pomer)

fig = go.Figure()
for key in results.keys():
    fig.add_trace(go.Scatter(x=results[key]['x'], 
                             y=results[key]['y'],
                             name=key,
                             mode='lines',
                             line=dict(color='rgba(255,0,0,1)', width=4) if key == 2002 else (dict(color='rgba(10,10,10,0.2)', width=2))
                             ))
    
fig.update_layout(
    title="Prírastok / úbytok v obyvateľstve podľa roku narodenia a veku",
    xaxis_title="vek",
    yaxis_title="prírastok / úbytok",
)
fig.show()

Na prvý pohľad vyzerá červená čiara (pomery počtu v nasledovných vekoch) úplne štandardne, ale potom si všimneme obrovský skok v deviatom roku života skupinky 2002, rovnako ako mierne väčší spád v 13-tom roku života. Stalo sa teda niečo špeciálne v 9. roku života ročníka 2002? 

#### Extrapolácia pokračovanie II.

Môžme si všimnúť, že ročník 2002 nie je jediný, ktorý takto poskočil v niektorom bode. Keď si myšou prejdeme všetky takéto kopce, všimneme si, že ročníky robiace tieto skoky idú po sebe a sú klesajúce. Kým ročník 2002 spravil vo svojich 9 rokoch veľký skok, ročník 2001 spravil vo svojich 10 rokoch veľký skor, ročník 2000 spravil vo svojich 11 rokoch veľký skok, atď. Vyzerá to na prvý pohľad, že v roku 2011 sa stalo niečo špeciálne. Skúsme si spraviť podobný graf, kde ale neporovnávame veky ale porovnávame priamo roky. Malá zmena kódu by nám mala pomôcť.

In [0]:
results = {}
for i_vek in range(1, len(relevant_data.index)):
    for i_rok in range(1, len(relevant_data.columns)):
        pomer = relevant_data.iloc[i_vek, i_rok] / relevant_data.iloc[i_vek-1, i_rok-1]
        narodeni = int(relevant_data.columns[i_rok]) - i_vek
        if narodeni not in results:
            results[narodeni] = {'x': [], 'y': []}
        results[narodeni]['x'].append(relevant_data.columns[i_rok]) # tu je zmena
        results[narodeni]['y'].append(pomer)

fig = go.Figure()
for key in results.keys():
    fig.add_trace(go.Scatter(x=results[key]['x'], 
                             y=results[key]['y'],
                             name=key,
                             mode='lines',
                             line=dict(color='rgba(255,0,0,1)', width=4) if key == 2002 else (dict(color='rgba(10,10,10,0.2)', width=2))
                             ))
fig.update_layout(
    title="Prírastok / úbytok v obyvateľstve podľa roku narodenia a roku",
    xaxis_title="rok",
    yaxis_title="prírastok / úbytok",
)
fig.show()

Teraz si môžme všimnúť, že špic v roku 2011 sa naozaj týka všetkých. Niektoré zmeny sú kladné a niektoré záporné. Ak si chvíľu budeme googliť, čo sa vtedy stalo, dostaneme sa k záveru, že v rokoch xxx1 sa robieva na Slovensku sčítanie obyvateľstva. Takže v týchto dátach vyzerajú byť v skutočnosti všetky dáta iba odhadované medzi rokmi končiacimi na 1. To nám trocha kazí dôveru v tento model, ale stále platí, že nemáme aktuálne nič lepšie.

Aby sme dokončili myšlienkový pochod, všimnime si, že úpadok v 13-tich rokoch je naozaj mierne výnimočný, pretože v tom roku ani v tom veku žiadna skupina nemala taký úpadok. Na tomto grafe vidíme špeciálny úpadok v prvom roku života, ale keď sa pozrieme na pôvodný graf, všimneme si, že takýto úpadok je celkom častý a teda by sme mu nemali prikladať špeciálny význam.

Došli sme teda k záveru, že náš ročník 2002 je z viacerých pohľadov v princípe normálnym ročníkom, až na mierny úpadok v roku 2015. 

### Zamyslenie a námety

Zdá sa vám, že sme ešte niečo nevzali v úvahu? Ak áno, určite sa nad tým zamyslite a skúste to vyhodnotiť.

Ukázaný postup pravdaže nie je jediný a budeme radi, ak nám dáte vedieť o iných zaujímavých postupoch, ktoré ste použili. V apríli sa možno dozvieme správne výsledky :)

Ak vás táto téma a dáta zaujali, budeme radi, ak sa skúsite zamyslieť a prediskutovať aj nasledovné otázky / prístupy:
- Čo by sa stalo, ak by sme sa snažili modelovať priamo prisťahovania, odsťahovania a úmrtia v jednotlivých vekoch, ako by sa líšili výsledky? Skúste nájsť relevantné dáta a použite ich.
- Ako súvisí istota vášho odhadu s vekom, ktorý sa snažíte odhadnúť? Je nejaký vek, pre ktorý by ste si svojim odhadom boli najistejší?
- Vedeli by ste podobnú analýzu zopakovať napríklad pre Ghanu?
- Predstavte si, že ste autobazár a rozhodujete sa, či spravíte reklamnú kampaň pre čerstvo plnoletých občanov. Ako by ste odhadli, aká časť ľudí dosahujúcich plnoletosť má na účte aspoň 3000 eur, za ktoré by si mohli kúpiť auto?

# Domáca úloha

Spoločnosť Anonymous International sa rozhodla, že o dva roky postaví nové obchodné centrum v Bratislave. V záujme šetrenia miesta sa spoločnosť rozhodla, že niektoré parkovacie miesta budú dĺžky 5m a niektoré budú dĺžky iba 4m.
1. Spravte analýzu, ktorá predpovie o dva roky pomer áut v Bratislave spadajúcich do týchto dvoch kategórií.
1. Anonymous International by chcela ísť s dobou a chceli by na tieto miesta zakomponovať aj miesta na nabíjanie elektromobilov. Koľko percent 4m, respektíve 5m áut budú o dva roky elektromobily?
